firedrake/parloops.py

```diff
@@ -103,11 +109,20 @@ def _form_loopy_kernel(kernel_domains, instructions, measure, args, **kwargs):
 
     if kernel_domains == "":
         kernel_domains = "[] -> {[]}"
-    kargs.append(...)
-    knl = loopy.make_function(kernel_domains, instructions, kargs, seq_dependencies=True,
-                              name="par_loop_kernel", silenced_warnings=["summing_if_branches_ops"])
-
-    return pyop2.Kernel(knl, "par_loop_kernel", **kwargs)
+    try:
+        key = (kernel_domains, tuple(instructions), tuple(map(tuple, kwargs.items())))
+        if kernel_cache is not None:
+            return kernel_cache[key]
+        else:
+            raise KeyError("No cache")
+    except KeyError:
+        kargs.append(...)
+        knl = loopy.make_function(kernel_domains, instructions, kargs, seq_dependencies=True,
+                                  name="par_loop_kernel", silenced_warnings=["summing_if_branches_ops"], target=loopy.CTarget())
+        if kernel_cache is not None:
+            return kernel_cache.setdefault(key, pyop2.Kernel(knl, "par_loop_kernel", **kwargs))
+        else:
+            return knl
```

Comment:
I screwed this bit up. I think it want's to be
```suggestion
        knl = pyop2.Kernel(knl, "par_loop_kernel", **kwargs)
        if kernel_cache is not None:
            return kernel_cache.setdefault(key, knl)
        else:
            return knl
```

Reply:
oh yes. fixed now.